### PR TITLE
feat: add citation sources to attachments

### LINKED ISSUE #85: Gemini Pro: support RECITATION finish reason
Gemini fails to answer questions which involve extensive verbatim citing of the data the model was originally trained on. 

Gemini stops streaming content with one of the following [finish reasons](https://github.com/googleapis/python-aiplatform/blob/9bda3288b59eb52c18a13c292561cb2c720ff331/google/cloud/aiplatform_v1beta1/types/content.py#L616-L651):

1. Successful generation:
    * FINISH_REASON_UNSPECIFIED 
    * STOP 
    * MAX_TOKENS 
2. Failed generation because of recitation:
   * RECITATION 
3. Failed generation because of unsafe content:
    * SAFETY 
    * BLOCKLIST 
    * PROHIBITED_CONTENT 
    * SPII 
4. Failed because of an unknown reason
    * OTHER  

The following queries typically fails because of recitation:

|Query|FinishReason|
|----|----|
|what are capitals of Latin America?|RECITATION, OTHER|
|recite the first chapter of the first Harry Potter book|RECITATION|
|count from 1 to 100|RECITATION|

When model fails wiht RECITATION, it provides a list of [citation sources](https://ai.google.dev/api/rest/v1/GenerateContentResponse#CitationMetadata) which could be displayed to the user. In fact, that's what Gemini playground chat does:

Screenshot 2024-05-14 at 11 16 15

Whereas the vertexai adapter reports `content_filter` finish reason when it sees RECITATION, which not exactly right. We should report the sources in attachments instead.

Note, that occasionally model can't find any sources at all:

Screenshot 2024-05-14 at 11 16 35

And it also could return OTHER finish reason:

Screenshot 2024-05-14 at 11 15 53

We could report the finish reason in `Info` stage.

### PR DESCRIPTION
**Pull request by roman-romanov-o**
resolves https://github.com/epam/ai-dial-adapter-vertexai/issues/85

### CODE REVIEW

File: aidial_adapter_vertexai/chat/gemini/adapter.py [hunk 1]
```diff
@@ -242,6 +245,28 @@ async def set_finish_reason(
...
+    if (
+        citation_metadata is None
+        or citation_metadata.citations is None
+        or not len(citation_metadata.citations)
+    ):
+        return None
+
+    for citation in citation_metadata.citations[:citations_limit]:
+        await consumer.add_attachment(
+            Attachment(url=citation.uri, title=citation.title)
```
Comment (roman-romanov-o) [inline aidial_adapter_vertexai/chat/gemini/adapter.py]: When I tested, this `citation.title` field was always empty

Maybe we should add some custom title like `Citation source`?

**Comment by roman-romanov-o** [comment]
/deploy-review
>- **Deployment failed**: [failed](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/483849)

File: aidial_adapter_vertexai/chat/gemini/adapter.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat/gemini/adapter.py]: No need to invent a title. If it's missing DIAL Chat prints the url itself, which is ok.

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-89.nightly-test.deltixhub.io
>E2E tests status: failed

File: aidial_adapter_vertexai/chat/gemini/adapter.py [hunk 2]
```diff
@@ -60,6 +60,8 @@
     HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
 }
 
+DEFAULT_CITATIONS_LIMIT = 10
```
Comment (adubovik) [inline aidial_adapter_vertexai/chat/gemini/adapter.py]: I don't think we need to hide anything from the user. Whatever the model returned, the user has a right get its full unabridged content.

File: aidial_adapter_vertexai/chat/gemini/adapter.py [hunk 3]
```diff
@@ -242,6 +245,28 @@ async def set_finish_reason(
...
+
+    if (
+        citation_metadata is None
+        or citation_metadata.citations is None
+        or not len(citation_metadata.citations)
+    ):
+        return None
+
+    for citation in citation_metadata.citations[:citations_limit]:
+        await consumer.add_attachment(
```
Comment (adubovik) [inline aidial_adapter_vertexai/chat/gemini/adapter.py]: Make sure that `citation.uri` isn't None too.
I see the following error:

```
pydantic.error_wrappers.ValidationError: 1 validation error for AttachmentChunk
__root__
  Trying to add attachment without data and url (type=value_error)
```

when I send the following request: `what are capitals of Latin America?`

Comment (roman-romanov-o) [reply aidial_adapter_vertexai/chat/gemini/adapter.py]: Yeah, interesting case
I haven't met this on just `gemini-pro`, but on `gemini1.5-pro` it reproduced

Somehow that citation had title, but uri was empty
Thanks for noticing!



Fixed

File: aidial_adapter_vertexai/chat/gemini/adapter.py (hunk 2 above)
Comment (roman-romanov-o) [reply aidial_adapter_vertexai/chat/gemini/adapter.py]: I was just sacred of 40 links that poped up on Harry Potter example :)

Fixed

**Comment by roman-romanov-o** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-89.nightly-test.deltixhub.io
>E2E tests status: failed

**Review by adubovik — APPROVED** [review]

**Pull request merged**